### ISSUE TITLE
Custom logit processor bypassed when speculative decoding is enabled

### ISSUE THREAD

**Issue by HH1162** [sgl-project/sglang#26330]
## Problem

When speculative decoding is enabled (e.g. `--speculative-algorithm NEXTN`), the custom logit processor registered via `--enable-custom-logit-processor` is not applied during token generation. The `thinking_budget` parameter in `custom_params` has no effect — the model generates reasoning tokens far beyond the configured budget.

Without speculative decoding, the same request with `thinking_budget: 1200` correctly limits reasoning to ~1200 tokens.

## Reproduction

**Server startup:**
```bash
sglang serve \
  --model-path /path/to/qwen3-fp8 \
  --trust-remote-code --quantization fp8 \
  --enable-custom-logit-processor \
  --reasoning-parser qwen3 \
  --speculative-algorithm NEXTN \
  --speculative-num-steps 5 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 6
```

**Request (curl, direct to SGLang):**
```bash
curl http://127.0.0.1:1234/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "qwen3",
    "messages": [{"role": "user", "content": "Prove that there are infinitely many primes using at least three different methods."}],
    "stream": true,
    "stream_options": {"include_usage": true},
    "custom_logit_processor": "{\"callable\": \"80049555000000000000008c2a73676c616e672e7372742e73616d706c696e672e637573746f6d5f6c6f6769745f70726f636573736f72948c225177656e33355468696e6b696e674275646765744c6f67697450726f636573736f729493942e\"}",
    "custom_params": {"thinking_budget": 1200}
  }'
```

The `custom_logit_processor` value is a hex-encoded pickle of:
```python
{"callable": "sglang.srt.sampling.custom_logit_processor.Qwen35ThinkingBudgetLogitProcessor"}
```

## Observed behavior

| Method | Speculative decoding | `reasoning_tokens` | `reasoning_chars` | Budget enforced? |
|--------|---------------------|-------------------|-------------------|------------------|
| curl (direct) | Disabled | ~1200 | ~3898 | Yes |
| curl (direct) | Enabled (NEXTN) | **~3848** | **~12352** | **No** |
| OpenAI SDK | Disabled | ~1200 | ~3898 | Yes |
| OpenAI SDK | Enabled (NEXTN) | **~2957-3112** | **~9281-11575** | **No** |

The HTTP request body was verified via client-side interception — `custom_logit_processor` and `custom_params` are present and correct in all cases. The server receives them but does not apply the logit processor when speculative decoding is active.

**Important:** This is not a token counting issue where draft tokens are counted but rejected. The actual `reasoning_chars` output (~9000-12000) far exceeds what 1200 tokens would produce (~3000-4000 chars), confirming that the logit processor is completely bypassed, not just miscounted.

## Expected behavior

The custom logit processor should be applied regardless of whether speculative decoding is enabled, enforcing the `thinking_budget` limit consistently.

## Root cause hypothesis

Speculative decoding generates draft tokens in batches (via the draft model) before verifying them with the target model. The custom logit processor is a per-token callback that modifies logits before sampling — it may be skipped during the draft phase, and the verification phase may not re-apply it, allowing budget-exceeding tokens through.

## Environment

- SGLang version: `0.5.13.dev20260525+ge86fdf3a3`
- Model: Qwen3 (FP8)
- GPU: RTX PRO 6000 Blackwell 96GB
- Python: 3.10

**Comment by HH1162** [comment]
Found related issue: #21724 which reports the same problem for EAGLE v2 speculative decoding. The root cause there was that the SPEC_V2 verify path skips `apply_custom_logit_processor`. 

Our case uses NEXTN speculative decoding (not EAGLE), and we do not have `SGLANG_ENABLE_SPEC_V2=1` set — yet the custom logit processor is still completely bypassed. This suggests the issue affects multiple speculative decoding algorithms, not just EAGLE v2.
